### PR TITLE
List idle users and periodic cleanup

### PR DESCRIPTION
**Pull request by tomtor**
Show users which never started a worker or created a test on the pending page.

Add periodic cleanup to scavenge.py

See https://github.com/glinscott/fishtest/pull/234

### CODE REVIEW

**Comment by Stefano80** [comment]
Sorry, @tomtor , I dont understand this. What is supposed to do?

**Comment by tomtor** [comment]
@Stefano80 See end of #234 comments

**Comment by ppigazzini** [comment]
server updated pulling this PR and the PR https://github.com/glinscott/fishtest/pull/241

**Comment by ppigazzini** [comment]
BTW I did a backup of the DB, so feel free to DROP ALL the Idle Users.

**Comment by tomtor** [comment]
@snicolet I suppose a periodic script is indeed the most obvious solution. It is easy to copy the python logic to a script like e.g. the scavenge.py.

@ppigazzini Shall I add it to scavenge.py?

**Comment by tomtor** [comment]
@ppigazzini moved to scavenge.py (now cleans idle tasks and 7 day idle users) and tested locally.

**Comment by snicolet** [comment]
@tomtor 
To clarify: we want to drop users which have registered more than 7 days ago, and have **never** submitted a test.

**Comment by ppigazzini** [comment]
dev-server updated, scavenge.py setup to run at every 5th minute, DB restore setup to run at every 5th minute from 2-59 minutes.

**Comment by ppigazzini** [comment]
@snicolet we need to keep also the pure CPU contributors

**Comment by ppigazzini** [comment]
@tomtor @snicolet I will prefer to wait 4 weeks before purging an Idle User: 4 weekends to try to setup a worker or to learn how to contribute a test. We need also to be sure to no drop an user that has contributed CPU for a 1000 games batch, or that dropping that user doesn't pollute the tests results.

**Comment by tomtor** [comment]
@ppigazzini What happens when you run 

python utils/scavenge.py

From the command line?

**Comment by ppigazzini** [comment]
@tomtor I view as output a long list of usernames, the Idle Users list is cleaned (there are only the users with registration date in the last 7 days)

File: fishtest/utils/scavenge.py [hunk 1]
```diff
@@ -19,8 +20,27 @@ def scavenge_tasks(scavenge=True, minutes=60):
...
+  for u in rundb.userdb.get_users():
+      if not 'registration_time' in u \
+         or u['registration_time'] < datetime.utcnow() - timedelta(days=days):
+        idle[u['username']] = u
+  for u in rundb.userdb.user_cache.find():
+    del idle[u['username']]
+  idle= idle.values()
+  return idle
+
+def scavenge_users(scavenge=True, days=7):
```
[the remark this answers is not in the document]
Comment (ppigazzini) [reply fishtest/utils/scavenge.py]: Wow, I didn't know that I can add a commit. So I should be able to solve the conflict of some PRs.

**Comment by Stefano80** [comment]
Hi @tomtor , it looks good to me too, just wen through the code.

**Comment by ppigazzini** [comment]
official server updated

**Comment by tomtor** [comment]
@ppigazzini I would have expected that on the official server the idle users without registration date would have been dropped by now?

**Comment by ppigazzini** [comment]
@tomtor We have done at the users the possibility to update the password, I'm wondering if a bad password can now break the DB.

**Comment by tomtor** [comment]
@ppigazzini what happens on the dev server when you replace the failing line with


    if u['username'] in idle:
      del idle[u['username']]
    else
      del idle[str(u['username'])]

**Comment by ppigazzini** [comment]
@tomtor (syntax error on else :D ) output w/ your changes
```
Scavenging {u'last_updated': datetime.datetime(2018, 4, 23, 8, 16, 15, 663000), u'worker_key': u'ChessDB_CN_WorkerPool-8cores', u'num_games': 1000, u'worker_info': {u'username': u'ChessDB_CN_WorkerPool', u'country_code': u'HK', u'unique_key': u'71a429d3-5b8e-4262-a26e-27f84130efeb', u'remote_addr': u'69.172.90.171', u'uname': u'Linux 3.10.0-693.21.1.el7.x86_64', u'version': 62, u'architecture': [u'64bit', u'ELF'], u'concurrency': 8}, u'residual': 1.0896555933662273, u'residual_color': u'#44EB44', u'pending': True, u'active': True, u'stats': {u'time_losses': 0, u'wins': 2, u'losses': 1, u'draws': 4, u'crashes': 0}, u'nps': 1574975.0}
Traceback (most recent call last):
  File "scavenge2.py", line 49, in <module>
    main()
  File "scavenge2.py", line 46, in main
    scavenge_users(scavenge=True)
  File "scavenge2.py", line 39, in scavenge_users
    for u in get_idle_users(days):
  File "scavenge2.py", line 34, in get_idle_users
    del idle[str(u['username'])]
KeyError: 'greentip1'
```


This is the last output of these changes
```
  for u in rundb.userdb.user_cache.find():
    print "user:"
    print u
    print u['username']
    print str(u['username'])
    print idle[str(u['username'])]
    print idle[u['username']]
    print "end"
    #del idle[u['username']]
  #idle= idle.values()
  return idle
```
```
user:
{u'username': u'greentip1', u'tests': 3, u'last_updated': u'Never', u'cpu_hours': 0, u'games': 0, u'games_per_hour': 0.0, u'_id': ObjectId('5add92db0ebc5972bf5de09b'), u'tests_repo': u'https://github.com/greentip1/Stockfish'}
greentip1
greentip1
Traceback (most recent call last):
  File "scavenge.py", line 55, in <module>
    main()
  File "scavenge.py", line 52, in main
    scavenge_users(scavenge=True)
  File "scavenge.py", line 44, in scavenge_users
    for u in get_idle_users(days):
  File "scavenge.py", line 35, in get_idle_users
    print idle[str(u['username'])]
KeyError: 'greentip1'
```

**Comment by ppigazzini** [comment]
@tomtor waiting for a patch I commented out `scavenge_users(scavenge=True)` on the official server.
Drop me a mail to <myname>_dot_<mysurname>_at_<gmail>_dot_<com> if you want the access to the dev server to debug the problem.
Pasquale

**Comment by tomtor** [comment]
This is very strange, we have a user greentip1which does not have an account but has created 3 tests?

Replacing the failing line with just

    if u['username'] in idle:
      del idle[u['username']]

Should prevent this runtime error without bad effects, but does not solve the riddle.

**Comment by ppigazzini** [comment]
On the dev server worked fine.
greentip1 seems to have joined 6 days ago : https://github.com/greentip1

We should check if there are other users in the same condition of greentip1.

**Comment by ppigazzini** [comment]
Here are the three tests submitted by greentip1:
http://tests.stockfishchess.org/tests/view/5ad9177f0ebc5902a3efed02
http://tests.stockfishchess.org/tests/view/5ad866340ebc595700526e95
http://tests.stockfishchess.org/tests/view/5ad755270ebc595700526e1e

**Comment by tomtor** [comment]
if u['username'] in idle:
      del idle[u['username']]
    else:
      print('Unkown user: ' + u['username'])

Should show these users.

**Comment by ppigazzini** [comment]
```
Unkown user: 8bittrex
Unkown user: greentip1
```

**Comment by tomtor** [comment]
Ahh, I get it now. Recent users (newer then the parameter n-days specified in the call to scavenge) are not placed in the initial idle list, so they cannot be removed. That's as it should be, so replacing the failing line by just:

    if u['username'] in idle:
      del idle[u['username']]

is the correct functional fix. @ppigazzini Could you make a commit for this?

**Comment by tomtor** [comment]
@ppigazzini I made a PR to fix it....

**Comment by tomtor** [comment]
@ppigazzini No all is well. Cleaning just aborted but nothing has been damaged on dev or production.

**Pull request merged**